File: Programmers/cpp/LEVEL_1_신고_결과_받기.cpp
#include <string>
#include <vector>
#include <set>
#include <unordered_map>

using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
	vector<int> answer;

	unordered_map<string, set<string>> userReportedList; //해당 유저를 신고한 유저 목록
	unordered_map<string, set<string>> userReportingList; // 각 유저가 신고한 목록
	set<string> reportList; //k번 이상 신고되어 정지될 유저

	for (string rep : report) {
		string s;
		string userId;
		string reportedId;

		for (int i = 0; i < rep.length(); i++) {
			if (rep[i] == ' ') {
				userId = s;
				s.clear();
				continue;
			}

			s += rep[i];
		}

		reportedId = s;

		userReportedList[reportedId].insert(userId);
		userReportingList[userId].insert(reportedId);
	}

	for (auto list : userReportedList) {
		if (list.second.size() >= k) {
			reportList.insert(list.first);
		}
	}

	for (string name : id_list) {
		set<string> reportSet = userReportingList[name];
		int cnt = 0;

		for (string rp : reportSet) {
			if (reportList.find(rp) != reportList.end()) {
				cnt++;
			}
		}

		answer.push_back(cnt);
	}

	return answer;
}

File: Programmers/java/LEVEL_1_신고_결과_받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];

        Map<String, Set<String>> userReportedList = new HashMap<>(); //해당 유저를 신고한 유저 목록
        Map<String, Set<String>> userReportingList = new HashMap<>(); //각 유저가 신고한 목록
        Set<String> reportList = new HashSet<>(); //k번 이상 신고되어 정지될 유저

        for (String rep : report) {
            String[] s = rep.split(" ");
            String userId = s[0];
            String reportedId = s[1];

            if (!userReportedList.containsKey(reportedId)) {
                userReportedList.put(reportedId, new HashSet<>());
            }

            userReportedList.get(reportedId).add(userId);

            if (!userReportingList.containsKey(userId)) {
                userReportingList.put(userId, new HashSet<>());
            }

            userReportingList.get(userId).add(reportedId);
        }

        for (String key : userReportedList.keySet()) {
            if (userReportedList.get(key).size() >= k) {
                reportList.add(key);
            }
        }

        int idx = 0;
        for (String name : id_list) {
            Set<String> reportSet = new HashSet<>(userReportingList.getOrDefault(name, new HashSet<>()));
            int cnt = 0;

            for (String rp : reportSet) {
                if (reportList.contains(rp)) {
                    cnt++;
                }
            }

            answer[idx] = cnt;
            idx++;
        }

        return answer;
    }
}